### PR TITLE
Changes to avoid gpstop errors when standby is not reachable.

### PR DESCRIPTION
**Pull request by Annu149**
Added check for standby reachability before stopping the standby to avoid gpstop failures when standby is unreachable.
Added check for unreachable hosts before cleaning shared memory at the end of gpstop process to avoid remove_shared_memory function from throwing errors when hosts are unreachable.
Added behave test case for gpstop when standby is unreachable.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gpstop [hunk 1]
```diff
@@ -321,16 +322,26 @@ class GpStop:
...
+        num_workers = min(len(self.gparray.get_hostlist()), self.parallel)
+        unreachable_hosts = get_unreachable_segment_hosts(cluster_hosts, num_workers)
+
         operations = []
         for gphost in cluster_hosts:
-            operations.append(RemoteOperation(CleanSharedMem(self.gparray.getDbList()), gphost))
+            if gphost not in unreachable_hosts:
+                operations.append(RemoteOperation(CleanSharedMem(self.gparray.getDbList()), gphost))
+            else:
+                logger.warning("Skipping cleaning shared memory on %s because it's unreachable" % (gphost))
```
Comment (shrakesh) [inline gpMgmt/bin/gpstop:333]: can we add or enhance the test to validate newly added warning for segment and standby.

File: gpMgmt/bin/gpstop [hunk 2]
```diff
@@ -568,6 +579,10 @@ class GpStop:
         if self.gparray.standbyMaster:
             standby = self.gparray.standbyMaster
 
+            if get_unreachable_segment_hosts([standby.hostname], 1):
+                logger.warning("Standby is unreachable, continuing to stop other segments...")
```
Comment (shrakesh) [inline gpMgmt/bin/gpstop:583]: can we have test to cover this warning as well.

File: gpMgmt/bin/gpstop (hunk 1 above)
Comment (Annu149) [reply gpMgmt/bin/gpstop:333]: sure, will add

File: gpMgmt/bin/gpstop (hunk 2 above)
Comment (Annu149) [reply gpMgmt/bin/gpstop:583]: sure, will cover this.

File: gpMgmt/bin/gpstop (hunk 1 above)
Comment (kaknikhil) [reply gpMgmt/bin/gpstop:333]: We might also be able to add unit tests to `test_unit_gpstop.py` for testing the new changes.

File: gpMgmt/bin/gpstop [hunk 3]
```diff
@@ -321,16 +322,26 @@ class GpStop:
...
             RemoteOperation(CleanSharedMem([self.gparray.master]), self.gparray.master.getSegmentHostName()))
 
         if self.gparray.standbyMaster:
-            operations.append(RemoteOperation(CleanSharedMem([self.gparray.standbyMaster]),
-                                              self.gparray.standbyMaster.getSegmentHostName()))
+            if self.gparray.standbyMaster.getSegmentHostName() not in unreachable_hosts:
+                operations.append(RemoteOperation(CleanSharedMem([self.gparray.standbyMaster]),
+                                                  self.gparray.standbyMaster.getSegmentHostName()))
+            else:
+                logger.warning("Skipping cleaning shared memory on %s because it's unreachable" % (
```
Comment (pobbatihari) [inline gpMgmt/bin/gpstop:343]: Can we improve log message.
for example :<Host> is unreachable. (Skipping/skip) cleaning shared memory.  you can think better

**Review by pobbatihari — APPROVED** [review]

File: gpMgmt/test/behave/mgmt_utils/steps/gpstart.py [hunk 4]
```diff
@@ -60,6 +60,25 @@ def cleanup(context):
 
     context.add_cleanup(cleanup, context)
 
+@when('the standby host is made reachable')
+@then('the standby host is made reachable')
+def impl(context):
```
Comment (divyeshddv) [inline gpMgmt/test/behave/mgmt_utils/steps/gpstart.py:65]: The cleanup step from the `the standby host is made unreachable` should ideally execute this at end of context/scenario. Maybe we can remove this step if thats the case (not needed)

And if we need it, we can also reuse the `cleanup()` defined in step above to reduce duplicate code

Comment (Annu149) [reply gpMgmt/test/behave/mgmt_utils/steps/gpstart.py:65]: We need "the standby host is made reachable" step , will use cleanup() function in this step to reduce duplicate code.

**Review by divyeshddv — APPROVED** [review]
LGTM!

**Pull request merged**